mapIt() C++ codegen breaks with HCR enabled

```nim
import std / sequtils

let x = [1]
echo(x.mapIt($it))
```
Compile with:
```
nim cpp --hotcodereloading mapit.nim
```
Gets:
```
/home/awr/.cache/nim/mapit_d/stdlib_system.nim.cpp: In function ‘void genericReset_actual(void*, TNimType*)’:
/home/awr/.cache/nim/mapit_d/stdlib_system.nim.cpp:1419:3: error: ‘genericResetAux_9bFXTAJ8caOL4Ioy2o3B3aw_actual’ was not declared in this scope
   genericResetAux_9bFXTAJ8caOL4Ioy2o3B3aw_actual(dest_0, (*mt_0).node);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/awr/.cache/nim/mapit_d/stdlib_system.nim.cpp:1419:3: note: suggested alternative: ‘genericResetAux_9bFXTAJ8caOL4Ioy2o3B3aw’
   genericResetAux_9bFXTAJ8caOL4Ioy2o3B3aw_actual(dest_0, (*mt_0).node);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   genericResetAux_9bFXTAJ8caOL4Ioy2o3B3aw
/home/awr/.cache/nim/mapit_d/stdlib_system.nim.cpp:1425:3: error: ‘genericResetAux_9bFXTAJ8caOL4Ioy2o3B3aw_actual’ was not declared in this scope
   genericResetAux_9bFXTAJ8caOL4Ioy2o3B3aw_actual(dest_0, (*mt_0).node);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/awr/.cache/nim/mapit_d/stdlib_system.nim.cpp:1425:3: note: suggested alternative: ‘genericResetAux_9bFXTAJ8caOL4Ioy2o3B3aw’
   genericResetAux_9bFXTAJ8caOL4Ioy2o3B3aw_actual(dest_0, (*mt_0).node);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   genericResetAux_9bFXTAJ8caOL4Ioy2o3B3aw
/home/awr/.cache/nim/mapit_d/stdlib_system.nim.cpp: In function ‘void genericAssignAux_69cRBG8C28ydeYlu49a9aOJCQ_actual(void*, void*, TNimNode*, bool)’:
/home/awr/.cache/nim/mapit_d/stdlib_system.nim.cpp:1542:3: error: ‘genericAssignAux_U5DxFPRpHCCZDKWQzM9adaw_actual’ was not declared in this scope
   genericAssignAux_U5DxFPRpHCCZDKWQzM9adaw_actual(((void*) ((NI)((NU64)(d) + (NU64)((*n_0).offset)))), ((void*) ((NI)((NU64)(s) + (NU64)((*n_0).offset)))), (*n_0).typ, 
shallow_0);                                                                                                                                                             
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/awr/.cache/nim/mapit_d/stdlib_system.nim.cpp:1542:3: note: suggested alternative: ‘genericAssignAux_U5DxFPRpHCCZDKWQzM9adaw’
   genericAssignAux_U5DxFPRpHCCZDKWQzM9adaw_actual(((void*) ((NI)((NU64)(d) + (NU64)((*n_0).offset)))), ((void*) ((NI)((NU64)(s) + (NU64)((*n_0).offset)))), (*n_0).typ, 
shallow_0);                                                                                                                                                             
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   genericAssignAux_U5DxFPRpHCCZDKWQzM9adaw
```

--- a/compiler/ccgcalls.nim
+++ b/compiler/ccgcalls.nim
@@ -184,8 +184,8 @@ template genParamLoop(params) {.dirty.} =
     if params != nil: add(params, ~", ")
     add(params, genArgNoParam(p, ri.sons[i]))
 
-proc addActualPrefixForHCR(res: var Rope, module: PSym, sym: PSym) =
-  if sym.flags * {sfImportc, sfNonReloadable} == {} and
+proc addActualSuffixForHCR(res: var Rope, module: PSym, sym: PSym) =
+  if sym.flags * {sfImportc, sfNonReloadable} == {} and sym.loc.k == locProc and
       (sym.typ.callConv == ccInline or sym.owner.id == module.id):
     res = res & "_actual".rope
 
@@ -203,7 +203,7 @@ proc genPrefixCall(p: BProc, le, ri: PNode, d: var TLoc) =
     genParamLoop(params)
   var callee = rdLoc(op)
   if p.hcrOn and ri.sons[0].kind == nkSym:
-    callee.addActualPrefixForHCR(p.module.module, ri.sons[0].sym)
+    callee.addActualSuffixForHCR(p.module.module, ri.sons[0].sym)
   fixupCall(p, le, ri, d, callee, params)
 
 proc genClosureCall(p: BProc, le, ri: PNode, d: var TLoc) =

--- a/compiler/ccgtypes.nim
+++ b/compiler/ccgtypes.nim
@@ -956,7 +956,7 @@ proc genProcHeader(m: BModule, prc: PSym, asPtr: bool = false): Rope =
   fillLoc(prc.loc, locProc, prc.ast[namePos], mangleName(m, prc), OnUnknown)
   genProcParams(m, prc.typ, rettype, params, check)
   # handle the 2 options for hotcodereloading codegen - function pointer
-  # (instead of forward declaration) or header for function budy with "_actual" postfix
+  # (instead of forward declaration) or header for function body with "_actual" postfix
   let asPtrStr = rope(if asPtr: "_PTR" else: "")
   var name = prc.loc.r
   if isReloadable(m, prc) and not asPtr:

--- a/compiler/cgen.nim
+++ b/compiler/cgen.nim
@@ -784,7 +784,7 @@ proc cgsym(m: BModule, name: string): Rope =
     rawMessage(m.config, errGenerated, "system module needs: " & name)
   result = sym.loc.r
   if m.hcrOn and sym != nil and sym.kind in {skProc..skIterator}:
-    result.addActualPrefixForHCR(m.module, sym)
+    result.addActualSuffixForHCR(m.module, sym)
 
 proc generateHeaders(m: BModule) =
   add(m.s[cfsHeaders], "\L#include \"nimbase.h\"\L")
@@ -1033,6 +1033,11 @@ proc genProcAux(m: BModule, prc: PSym) =
     generatedProc = ropecg(p.module, "$N$1 {$n$2$3$4}$N$N",
                          [header, p.s(cpsLocals), p.s(cpsInit), p.s(cpsStmts)])
   else:
+    if m.hcrOn and isReloadable(m, prc):
+      # Add forward declaration for "_actual"-suffixed functions defined in the same module (or inline).
+      # This fixes the use of methods and also the case when 2 functions within the same module
+      # call each other using directly the "_actual" versions (an optimization) - see issue #11608
+      addf(m.s[cfsProcHeaders], "$1;\n", [header])
     generatedProc = ropecg(p.module, "$N$1 {$N", [header])
     add(generatedProc, initGCFrame(p))
     if optStackTrace in prc.options:

--- a/lib/nimhcr.nim
+++ b/lib/nimhcr.nim
@@ -421,7 +421,7 @@ when defined(createNimHcr):
       modules.add(name, newModuleDesc())
 
     let copiedName = name & ".copy." & dllExt
-    copyFile(name, copiedName)
+    copyFileWithPermissions(name, copiedName)
 
     let lib = loadLib(copiedName)
     assert lib != nil

